lib/mail_catcher/smtp.rb

```diff
@@ -1,58 +1,415 @@
+                write_response 250, 'OK'
+              when 'NOOP'
+                if line && !line.empty?
+                  write_response 501, 'Unexpected parameters or arguments'
+                  next
+                end
+                write_response 250, 'OK'
+              when 'EXPN'
+                write_response 502, 'Command not implemented'
+              when 'VRFY'
+                write_response 502, 'Command not implemented'
+              when 'HELP'
+                write_response 502, 'Command not implemented'
+              when 'STARTTLS'
+                write_response 502, 'Command not implemented'
+              when 'QUIT'
+                if line && !line.empty?
+                  write_response 501, 'Unexpected parameters or arguments'
+                  next
+                end
+                write_response 221, 'Bye!'
+                close unless closed?
+                return
+              else
+                write_response 500, 'Syntax error, command unrecognized'
+              end
+
+              task.yield
+            end
+
+            close unless closed?
+          end
+        end
+      end
     end
-    true
-  end
 
-  def receive_message
-    MailCatcher::Mail.add_message current_message
-    MailCatcher::Mail.delete_older_messages!
-    puts "==> SMTP: Received message from '#{current_message[:sender]}' (#{current_message[:source].length} bytes)"
-    true
-  rescue => exception
-    MailCatcher.log_exception("Error receiving message", @current_message, exception)
-    false
-  ensure
-    @current_message = nil
+    require 'async'
+    require 'async/task'
+    require 'async/io/stream'
+
+    class Server
+      def initialize(endpoint, protocol = endpoint.protocol)
+        @endpoint = endpoint
+        @protocol = protocol
+
+        define_singleton_method(:call, &proc) if block_given?
+      end
+
+      def accept(peer, address, task: Async::Task.current)
+        Async.logger.debug(self) { "Incoming connnection from #{address.inspect}" }
+
+        stream = Async::IO::Stream.new(peer)
+        protocol = @protocol.server(stream, hostname: @endpoint.hostname)
+
+        protocol.each do |envelope|
+          Async.logger.debug(self) { "Incoming message from #{address.inspect}: #{envelope.inspect}" }
+
+          call(envelope)
+        end
+
+        Async.logger.debug(self) { "Connection from #{address.inspect} closed cleanly" }
+      rescue EOFError, Errno::ECONNRESET, Errno::EPIPE, Errno::EPROTOTYPE
+        # Sometimes client will disconnect without completing a result or reading the entire buffer. That means we are done.
+        # Errno::EPROTOTYPE is a bug with Darwin. It happens because the socket is lazily created (in Darwin).
+        Async.logger.debug(self) { "Connection from #{address.inspect} closed: #{$ERROR_INFO}" }
+      end
+
+      def call
+        raise NotImplementedError, 'Supply a block to MailCatcher::SMTP::Server.new or subclass and implement #call'
```

Comment:
I'm not a contributor so I'm just chiming in to say that `NotImplementedError` is meant to indicate that a feature is not implemented on the current platform (e.g. mre or jruby) and that it is not caught by `begin..rescue` unless explicitly specified in the `rescue` clause.
In order to warn the user a `NoMethodError` can be raised in the same way or any exception that descends from `StandardError`